spynnaker/pyNN/external_devices_models/spif_input_device.py

```diff
@@ -0,0 +1,296 @@
+        """
+        return FPGAConnection(
+            SPIF_FPGA_ID, SPIF_OUTPUT_FPGA_LINK, board_address, chip_coords)
+
+    def __fpga_index(self, fpga_link_id):
+        # We use every other odd link, so we can work out the "index" of the
+        # link in the list as follows
+        return (fpga_link_id - 1) // 2
+
+    @overrides(ApplicationFPGAVertex.get_incoming_slice_for_link)
+    def get_incoming_slice_for_link(self, link, index):
+        vertex_slice = super().get_incoming_slice_for_link(link, index)
+        self.__index_by_slice[link.fpga_link_id, vertex_slice] = index
+        return vertex_slice
+
+    @overrides(ApplicationFPGAVertex.get_machine_fixed_key_and_mask)
+    def get_machine_fixed_key_and_mask(self, machine_vertex, partition_id):
+        fpga_link_id = machine_vertex.fpga_link_id
+        vertex_slice = machine_vertex.vertex_slice
+        index = self.__index_by_slice[fpga_link_id, vertex_slice]
+        key = ((self.__base_key << self.__neuron_bits) |
+               index << self.__index_shift)
+
+        fpga = self.__fpga_index(fpga_link_id)
+
+        # Build the key from the components
+        fpga_key = key + fpga
+        fpga_mask = self.__m_vertex_mask | self.__spif_mask
+        return BaseKeyAndMask(fpga_key, fpga_mask)
+
+    @overrides(ApplicationFPGAVertex.get_fixed_key_and_mask)
+    def get_fixed_key_and_mask(self, partition_id):
+        n_key_bits = BITS_IN_KEY - self.__neuron_bits
+        key_mask = ((1 << n_key_bits) - 1) << self.__neuron_bits
+        return BaseKeyAndMask(self.__base_key << self.__neuron_bits, key_mask)
+
+    @property
+    @overrides(AbstractSendMeMulticastCommandsVertex.start_resume_commands)
+    def start_resume_commands(self):
+        # Make sure everything has stopped
+        commands = [SpiNNFPGARegister.STOP.cmd()]
+
+        # Clear the counters
+        commands.append(SPIFRegister.OUT_PERIPH_PKT_CNT.cmd(0))
+        commands.append(SPIFRegister.CONFIG_PKT_CNT.cmd(0))
+        commands.append(SPIFRegister.DROPPED_PKT_CNT.cmd(0))
+        commands.append(SPIFRegister.IN_PERIPH_PKT_CNT.cmd(0))
+
+        # Configure the creation of packets from fields to keys using the
+        # "standard" input to SPIF (X | P | Y) and convert to (Y | X)
+        commands.extend([
+            set_field_mask(self.__pipe, 0, self.__input_mask),
+            set_field_shift(self.__pipe, 0, self.__input_shift),
+            set_field_limit(self.__pipe, 0, self.n_atoms),
+            # These are unused but set them to be sure
+            set_field_mask(self.__pipe, 1, 0),
+            set_field_shift(self.__pipe, 1, 0),
+            set_field_limit(self.__pipe, 1, 0),
+            set_field_mask(self.__pipe, 2, 0),
+            set_field_shift(self.__pipe, 2, 0),
+            set_field_limit(self.__pipe, 2, 0),
+            set_field_mask(self.__pipe, 3, 0),
+            set_field_shift(self.__pipe, 3, 0),
+            set_field_limit(self.__pipe, 3, 0)
+        ])
```

Comment:
Magic numbers 0, 1, 2, 3
a comment on what they are would be ok

Reply:
Replaced with a loop here over the things not used